### PR TITLE
luminous: osd/ReplicatedBackend.cc: 1349: FAILED ceph_assert(peer_missing.count(fromshard))

### PR DESCRIPTION
**Pull request by smithfarm**
backport tracker: https://tracker.ceph.com/issues/41730

---

backport of:

- [x] https://github.com/ceph/ceph/pull/30119 (merged September 4, 2019)
- [x] https://github.com/ceph/ceph/pull/30059 (merged September 7, 2019)
- [x] https://github.com/ceph/ceph/pull/30226 (merged September 9, 2019)

parent tracker: https://tracker.ceph.com/issues/41385

this backport was staged using ceph-backport.sh version 15.0.0.6950
find the latest version at https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

**Comment by smithfarm** [comment]
@neha-ojha Not clear to me how to backport #30059 to luminous. Either you could advise me how to do it, or close this PR and take it over?

**Comment by neha-ojha** [comment]
> @neha-ojha Not clear to me how to backport #30059 to luminous. Either you could advise me how to do it, or close this PR and take it over?

Since https://github.com/ceph/ceph/commit/d2ca3d2feb442f97ca89023c7d01178d96f517a6#diff-dfb9ddca0a3ee32b266623e8fa489626R2550 wasn't backported to luminous, I think we should do the following:

`diff --git a/src/osd/PG.cc b/src/osd/PG.cc`
`index aae386c..07e4c99 100644`
`--- a/src/osd/PG.cc`
`+++ b/src/osd/PG.cc`
`@@ -2321,10 +2321,13 @@ void PG::finish_recovery(list<Context*>& tfin)`
` void PG::_finish_recovery(Context *c)`
` {`
 `  lock();`
`-  if (deleting) {`
`+  if (deleting || !is_clean()) {`
`+    dout(10) << __func__ << " raced with delete or repair" << dendl;`
 `    unlock();`
    ` return;`
`   }`
`+  // When recovery is initiated by a repair, that flag is left on`
`+  state_clear(PG_STATE_REPAIR);`
  ` if (c == finish_sync_event) {`
   `  dout(10) << "_finish_recovery" << dendl;`
 `    finish_sync_event = 0;`




 @dzafman what do you think?

**Comment by dzafman** [comment]
@smithfarm @neha-ojha 

Just do this. 

> `diff --git a/src/osd/PG.cc b/src/osd/PG.cc`
> `index aae386c..07e4c99 100644`
> `--- a/src/osd/PG.cc`
> `+++ b/src/osd/PG.cc`
> `@@ -2321,10 +2321,13 @@ void PG::finish_recovery(list<Context*>& tfin)`
> ` void PG::_finish_recovery(Context *c)`
> ` {`
> ` lock();`
> `- if (deleting) {`
> `+ if (deleting || !is_clean()) {`
> `+ dout(10) << __func__ << " raced with delete or repair" << dendl;`
> ` unlock();`
> ` return;`
> ` }`

 Don't add the comment nor state_clear(PG_STATE_REPAIR);.  Those lines were moved but don't exist in Luminous.

**Comment by smithfarm** [comment]
if it passes "make check", the DNM could be removed and the PR could be tagged for integration testing

**Comment by smithfarm** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc: In member function 'void PG::MissingLoc::remove_stray_recovery_sources(pg_shard_t)':
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc:800:5: error: 'now_stray' was not declared in this scope
     now_stray.insert(*p);
     ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.cc:804:7: error: 'now_stray' was not declared in this scope
   if (now_stray.empty()) {
       ^
src/osd/CMakeFiles/osd.dir/build.make:101: recipe for target 'src/osd/CMakeFiles/osd.dir/PG.cc.o' failed
make[3]: *** [src/osd/CMakeFiles/osd.dir/PG.cc.o] Error 1
```

File: src/osd/PG.cc [hunk 1]
```diff
@@ -759,7 +759,56 @@ void PG::MissingLoc::check_recovery_sources(const OSDMapRef& osdmap)
...
+  set<pg_shard_t> now_stray;
+  for (set<pg_shard_t>::iterator p = missing_loc_sources.begin();
+       p != missing_loc_sources.end();
+       ) {
+    if (*p != stray) {
+      ++p;
+      continue;
+    }
+    ldout(pg->cct, 10) << __func__ << " source osd." << *p << " now stray" << dendl;
+    now_stray.insert(*p);
```
Comment (smithfarm) [inline src/osd/PG.cc]: @neha-ojha This `now_stray` is not defined in luminous... :-(

Comment (dzafman) [reply src/osd/PG.cc]: now_stray was removed by "osd/MissingLoc.cc: do not rely on missing_loc_sources only" so the manual backport of 5906a57320f04f57a38eef9588bd16ac3fd4e55d isn't quite right.

Comment (smithfarm) [reply src/osd/PG.cc]: Thanks, @dzafman - fixed now!

File: src/osd/PG.cc [hunk 2]
```diff
@@ -759,7 +759,48 @@ void PG::MissingLoc::check_recovery_sources(const OSDMapRef& osdmap)
...
+  }
+  // filter missing_loc_sources
+  for (set<pg_shard_t>::iterator p = missing_loc_sources.begin();
+       p != missing_loc_sources.end();
+       ) {
+    if (*p != stray) {
+      ++p;
+      continue;
+    }
+    ldout(cct, 10) << __func__ << " remove osd" << stray << " from missing_loc_sources" << dendl;
```
Comment (dzafman) [inline src/osd/PG.cc]: Need pg->cct

**Comment by smithfarm** [comment]
@dzafman Thanks for your help. Should be good to go now?

**Comment by yuriw** [comment]
[wip-yuri3-testing-2019-12-23-1538-luminous](https://trello.com/c/s72CrBPs/788-wip-yuri3-testing-2019-12-23-1538-luminous)

**Pull request merged**